dora/core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java

```diff
@@ -138,49 +137,52 @@ protected Key generateKey(AlluxioURI ufsUri, UnderFileSystemConfiguration ufsCon
   private UnderFileSystem getOrAddWithRecorder(AlluxioURI ufsUri,
       UnderFileSystemConfiguration ufsConf, Recorder recorder) {
     Key key = generateKey(ufsUri, ufsConf);
+    return get(key)
+        .orElseGet(() -> add(ufsUri, ufsConf, recorder));
+  }
+
+  protected Key generateKey(AlluxioURI ufsUri, UnderFileSystemConfiguration ufsConf) {
+    return new Key(ufsUri, ufsConf.getMountSpecificConf());
+  }
+
+  // On cache miss, synchronize the creation to ensure ufs is only created once
+  private synchronized UnderFileSystem add(AlluxioURI ufsUri, UnderFileSystemConfiguration ufsConf,
+      Recorder recorder) {
+    Key key = generateKey(ufsUri, ufsConf);
```

Comment:
This function is called by getOrAdd() and getOrAddWithRecorder().
Those two functions have already called the generateKey(). Now the generateKey() is called again in this function. This is duplicated? Pass the result `key` to this function as a parameter?